openrtb-core/src/main/protobuf/openrtb.proto

```diff
@@ -1197,6 +1203,78 @@ message BidRequest {
+  // device.sua should be considered the more accurate representation of the device attributes.
+  // This is because the device.ua may contain a frozen or reduced user agent string
+  // due to deprecation of user agent strings by browsers.
+  message UserAgent {
+    // Each BrandVersion object identifies a browser
+    // or similar software component. Implementers should send brands and
+    // versions derived from the Sec-CH-UA-Full-Version-List header
+    // or an equivalent JavaScript accessor from NavigatorUAData interface.
+    // This header or accessor are only available for browsers that support User-Agent Client Hints.
+    // RECOMMENDED by the OpenRTB specification.
+    repeated BrandVersion browsers = 1;
+
+    // A BrandVersion object that identifies the
+    // user agent’s execution platform / OS. Implementers should send a brand
+    // derived from the Sec-CH-UA-Platform header, and version derived
+    // from the Sec-CH-UAPlatform-Version header
+    // or an equivalent JavaScript accessor from NavigatorUAData interface.
+    // This header or accessor are only available for browsers that support User-Agent Client Hints.
+    // RECOMMENDED by the OpenRTB specification.
+    optional BrandVersion platform = 2;
+
+    // 1 if the agent prefers a “mobile” version of the content, if available,
+    // i.e. optimized for small screens or touch input.
+    // 0 if the agent prefers the “desktop” or “full” content.
+    // Implementers should derive this value from the Sec-CH-UAMobile header
+    // or an equivalent JavaScript accessor from NavigatorUAData interface.
+    // This header or accessor are only available for browsers that support User-Agent Client Hints.
+    optional bool mobile = 3;
+
+    // Device’s major binary architecture, e.g. “x86” or “arm”.
+    // Implementers should retrieve this value from the Sec-CH-UA-Arch header
+    // or an equivalent JavaScript accessor from NavigatorUAData interface.
+    // This header or accessor are only available for browsers that support User-Agent Client Hints.
+    optional string architecture = 4;
+
+    // Device’s bitness, e.g. “64” for 64-bit architecture.
+    // Implementers should retrieve this value from the Sec-CH-UA-Bitness header
+    // or an equivalent JavaScript accessor from NavigatorUAData interface.
+    // This header or accessor are only available for browsers that support User-Agent Client Hints.
+    optional string bitness = 5;
+
+    // Device model. Implementers should retrieve this value from the Sec-CH-UAModel header
+    // or an equivalent JavaScript accessor from NavigatorUAData interface.
+    // This header or accessor are only available for browsers that support User-Agent Client Hints.
+    optional string model = 6;
+
+    // The source of data used to create this object, List: User-Agent Source in AdCOM 1.0
+    optional int32 source = 7 [default = 0];
```

Comment:
should this be using the `UserAgentSource` enum instead ?

Reply:
Is not using enums within protobufs a general DV+ practice ?
https://developers.google.com/authorized-buyers/rtb/downloads/openrtb-proto#:~:text=//%20Possible%20sources%20for,3%3B%0A%7D
Enum is in the IAB openrtb proto file

Reply:
it is generally avoided when the schema is shared between two different applications

Reason being, if the producer adds a new enum value and starts sending it, consumer validations of the message might break till the consumer app switches to using the new schema too. Therefore making changes to the enum becomes troublesome. Using an integer helps work around this issue.

Reply:
Yes, we don't use enums in the proto schema, but the comment should refer to the enum it represents, e.g for `at`: https://github.com/IABTechLab/openrtb-proto-v2/blob/master/openrtb-core/src/main/protobuf/openrtb.proto#L72

Lots of other examples too

Reply:
Thanks @mforte-magnite, good question. The team is correct. The enums are still provided as a convenience, for code readability. The primary reason for not using them directly here is that OpenRTB enums are generally defined as extensible -- usually values 500 and up are available for customization. Proto deserialization then causes the problem that @cindyhere33 references.

Reply:
Added enum definition in https://github.com/IABTechLab/adcom-proto/pull/15